criu/file-ids.c

```diff
@@ -77,11 +78,18 @@ int fd_id_generate_special(struct fd_parms *p, u32 *id)
 {
 	if (p) {
 		struct fd_id *fi;
+		struct stat st_kfd;
 
 		fi = fd_id_cache_lookup(p);
 		if (fi) {
-			*id = fi->id;
-			return 0;
+			if (stat(AMDGPU_KFD_DEVICE, &st_kfd) == -1) {
```

Comment:
This looks wrong. If /dev/kfd exists, criu doesn't cache file fd-s.... Do I miss something?

Reply:
Yes, we don't want criu to cache the fds for kfd processes. I've put some more details in the commit message.

Reply:
but you disable the cache for all files if /dev/kfd exits and I haven't found any explanation why we need to do this in the commit message.

Maybe you want to disable the cache only for /dev/kfd files?

Reply:
I will update the PR with the suggested changes.

Reply:
Hi @avagin - Yes, I wanted to just disable caching for /dev/kfd (amdgpu devices). I spent some more time debugging this but so far I haven't figured out a clean way to deal with this. I was thinking of using HANDLE_DEVICE_VMA plugin which I introduced in the first patch of the series to also detect whether we don't want the id caching during checkpoint and during the last stage of checkpoint i.e. cr_dump_finish enable it again but thats not a good solution either. Is there a cleaner way of disabling caching of ids on the fly and then enable again? Maybe a new command line param if thats acceptable? Also, I am trying to figure out if this is a generic issue with tasks that have device file fds containing shared memory / anonymous memory mappings.  Please advice.

Reply:
```
diff --git a/criu/file-ids.c b/criu/file-ids.c
index 1b9d68888..7a6e6b5b6 100644
--- a/criu/file-ids.c
+++ b/criu/file-ids.c
@@ -72,6 +72,10 @@ static struct fd_id *fd_id_cache_lookup(struct fd_parms *p)
 
 int fd_id_generate_special(struct fd_parms *p, u32 *id)
 {
+       // Don't cache mapped devices.
+       if (p->st_mode & (S_IFCHR|S_IFBLK))
+               p = NULL;
+
        if (p) {
                struct fd_id *fi;
```

I think we can skip all devices here. fd_id_cache_lookup is used for memory mappings and special files like cwd, exe, etc.

Reply:
In an ideal case, we have to be able to get a file descriptor that is associated with a mapped struct file. In this case, we will be able to use kcmp and find out all mappings and file descriptors that are shared the same struct file. But this will require some kernel changes...

Reply:
Thanks @avagin .. I have uploaded a slightly modified version otherwise I was getting a seg fault while testing.